subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DuplicateHandlingCopyActionDecorator.java

```diff
@@ -66,6 +67,30 @@ public WorkResult execute(final CopyActionProcessingStream stream) {
         }));
     }
 
+    private String explainDuplicateLocation(FileCopyDetailsInternal copyDetails) {
+        if (copyDetails.isCompressedArchiveEntry()) {
+            return String.format("\"%s\", extracted from \"%s\",", copyDetails.getRelativePath(), extractCompressedFileName(copyDetails));
+        } else {
+            return String.format("\"%s\"", copyDetails.getRelativePath());
+        }
+    }
+
+    private String extractCompressedFileName(FileCopyDetailsInternal copyDetails) {
+        String displayName = FilenameUtils.separatorsToUnix(copyDetails.getDisplayName());
+        int exclamationIndex = displayName.indexOf("!");
+        if (exclamationIndex == -1) {
+            throw new IllegalStateException("Expected to find '!' in display name of compressed file entry: " + displayName);
+        }
+
+        String pathBeforeExclamation = displayName.substring(0, exclamationIndex);
+        int lastSlashIndex = pathBeforeExclamation.lastIndexOf("/");
+        if (lastSlashIndex == -1) {
+            throw new IllegalStateException("Expected to find '/' in display name of compressed file entry: " + displayName);
+        }
+
+        return pathBeforeExclamation.substring(lastSlashIndex + 1);
+    }
```

Comment:
❌ This isn't what we should do. 

The wording isn't very good and I think focuses on the wrong things. It looks like it's saying there are duplicates in the archive file.

This is the message we produce with the PR now:
> Encountered duplicate path "sub/c.txt", extracted from "files.zip", during copy operation configured with DuplicatesStrategy.FAIL

I think we want to say something like:
> Cannot copy <zip entry> because it has already been copied from <previous source>

IMO, the detail about duplicates strategy isn't useful but we could add "and the copy is configured to not allow duplicates".